File: cpu-bound/nodejs/benchmark.js
const { PerformanceObserver, performance } = require('perf_hooks');
const { computation, times } = require('./computation')
const { fork } = require('child_process')
const { Worker } = require('worker_threads')

const observer = new PerformanceObserver((items) => {
  items.getEntries().forEach((entry) => {
    console.log(`${entry.name}: ${entry.duration / 1000.0} seconds`)
  })
  performance.clearMarks();
});

observer.observe({ type: 'measure' });
performance.measure('start')

performance.mark('sync');

times().forEach(() => {
  computation()
})

performance.measure('sync', 'sync');

performance.mark('forking');

times().forEach(() => {
  const child = fork('./cpu-bound/nodejs/child-process.js')
})

performance.measure('forking', 'forking');

performance.mark('workerthreads');

times().forEach(() => {
  const worker = new Worker('./cpu-bound/nodejs/worker-thread.js', {})
})

performance.measure('workerthreads', 'workerthreads');

File: cpu-bound/benchmark.rb
require 'benchmark'
require './cpu-bound/computation'

def sync; times.times { computation }; end

def threads
  times.times.map { Thread.new { computation }}.each(&:join)
end

def fibers
  times.times.map { Fiber.new { computation }}.each(&:resume)
end

def forking
  times.times.map { fork { computation }}; Process.waitall
end

def ractors
  times.times.map { Ractor.new { computation }}.each(&:take)
end

Benchmark.bm do |x|
  x.report('sync')    {  sync    }
  x.report('threads') {  threads }
  x.report('fibers')  {  fibers  }
  x.report('forking') {  forking }
  x.report('ractors') {  ractors }
end
